tests/testautonatservice.nim

```diff
@@ -0,0 +1,150 @@
+# Nim-LibP2P
+# Copyright (c) 2022 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import std/options
+import chronos
+import unittest2
+import ../libp2p/[builders,
+                  switch,
+                  services/autonatservice]
+import ./helpers
+import stubs/autonatstub
+
+proc createSwitch(): Switch =
+  var builder = SwitchBuilder.new()
+    .withRng(newRng())
+    .withAddresses(@[ MultiAddress.init("/ip4/0.0.0.0/tcp/0").tryGet() ])
+    .withTcpTransport()
+    .withMplex()
+    .withAutonat()
+    .withNoise()
+
+  return builder.build()
+
+suite "Autonat Service":
+  teardown:
+    checkTrackers()
+
+  asyncTest "Autonat Service Private Reachability test":
+
+    let switch1 = createSwitch()
+    let switch2 = createSwitch()
+    let switch3 = createSwitch()
+    let switch4 = createSwitch()
+
+    let autonatStub = AutonatStub.new(expectedDials = 3)
+    autonatStub.returnSuccess = false
+
+    let autonatService = AutonatService.new(autonatStub, newRng(), some(1.seconds))
+    switch1.addService(autonatService)
```

Comment:
not related per se to the test, but does it make sense to pass the `autonatService` in the switch builder `.withAutonat()`?

So somthing like `.withAutonat(autonatService)` so that we avoid the calling `addService`?

Reply:
Good point, thanks for reminding me about this. Autonat and AutonatService are different things. We are planning to implement clearer abstractions soon. For now, perhaps something like `withServices(...)`?

Reply:
ah right. `withServices()` sounds good, but totally up to you :)
nothing against the current implementation though then.